### ISSUE TITLE
Runnning the ultralytics/examples/YOLOv8-CPP-inference code. but the detected picture has no box(All box outputs are (0,0,0,0))

### ISSUE THREAD

**Issue by all-for-code** [ultralytics/ultralytics#6597]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Predict

### Bug

code output:
Running on CUDA
Number of detections:1
Number of detections:1

I added the print information in inference.cpp and the output is as follows：
![image](https://github.com/ultralytics/ultralytics/assets/110803332/9b229719-7ed4-46d4-bf6e-c71598f02374)
output:
Running on CUDA
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
Number of detections:1
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
Number of detections:1
which means all detected boxes are (0,0,0,0)

### Environment

Ultralytics YOLOv8.0.218 🚀 Python-3.10.12 torch-2.1.1+cu121 CUDA:0 (NVIDIA GeForce RTX 3090, 24260MiB)
Setup complete ✅ (40 CPUs, 125.5 GB RAM, 801.8/878.6 GB disk)

OS                  Linux-6.2.0-36-generic-x86_64-with-glibc2.35
Environment         Linux
Python              3.10.12
Install             pip
RAM                 125.54 GB
CPU                 Intel Xeon Silver 4210 2.20GHz
CUDA                12.1
opencv            4.8
cmake             3.22.1

### Minimal Reproducible Example

I am using the official code, except for changing the shape of the input image to (640,640) and adding some print information, there are no other changes。
![image](https://github.com/ultralytics/ultralytics/assets/110803332/7ec4aada-8dea-4ff4-80ba-369ad95ab4c9)

### Additional

The following is my specific use of the code steps：
1. I used an image provided by nvidia：$ docker pull nvidia/cuda:12.0.0-devel-ubuntu22.04
2. Compile opencv4.8 using the build_opencv.sh script: 
`VERSION=4.8.0

test -e ${VERSION}.zip || wget https://github.com/opencv/opencv/archive/refs/tags/${VERSION}.zip
test -e opencv-${VERSION} || unzip ${VERSION}.zip

test -e opencv_extra_${VERSION}.zip || wget -O opencv_extra_${VERSION}.zip https://github.com/opencv/opencv_contrib/archive/refs/tags/${VERSION}.zip
test -e opencv_contrib-${VERSION} || unzip opencv_extra_${VERSION}.zip

cd opencv-${VERSION}
mkdir build
cd build

cmake -D CMAKE_BUILD_TYPE=RELEASE \
-D CMAKE_INSTALL_PREFIX=/usr/local \
-D WITH_TBB=ON \
-D ENABLE_FAST_MATH=1 \
-D CUDA_FAST_MATH=1 \
-D WITH_CUBLAS=1 \
-D WITH_CUDA=ON \
-D BUILD_opencv_cudacodec=ON \
-D WITH_CUDNN=ON \
-D OPENCV_DNN_CUDA=ON \
-D WITH_QT=OFF \
-D WITH_OPENGL=ON \
-D BUILD_opencv_apps=OFF \
-D BUILD_opencv_python2=OFF \
-D OPENCV_GENERATE_PKGCONFIG=ON \
-D OPENCV_PC_FILE_NAME=opencv.pc \
-D OPENCV_ENABLE_NONFREE=ON \
-D OPENCV_EXTRA_MODULES_PATH=../../opencv_contrib-${VERSION}/modules \
-D INSTALL_PYTHON_EXAMPLES=OFF \
-D INSTALL_C_EXAMPLES=OFF \
-D BUILD_EXAMPLES=OFF \
-D CUDA_ARCH_BIN=8.6 \
-D WITH_FFMPEG=ON \
-D CUDNN_INCLUDE_DIR=/usr/local/cuda/include \
-D CUDNN_LIBRARY=/usr/local/cuda/lib64/libcudnn.so \
..

make -j 8
sudo make -j 8 install`

4. I change the input of the image to (640,640) and the path of onnx file which is created by the code:
`from ultralytics import YOLO

model = YOLO("yolov8_pt/yolov8s.pt")  # load a pretrained model (recommended for training)

model.export(format="onnx", device=3, imgsz = (640,640), half=True, batch=1, dynamic=False)  #  export the model to ONNX format
`
the yolov8s.pt is download from the link: [https://github.com/ultralytics/assets/releases](url)
5. cd ultralytics/examples/YOLOv8-CPP-inference; mkdir build;  cd build; cmake ..; make -j8; ./Yolov8CPPInference

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@all-for-code hi there,

Thanks for reaching out with your issue. Based on the details you've provided, it seems like the exported ONNX model may not be providing the right detection outputs, with all bounding boxes being (0,0,0,0). This could be due to various reasons, such as issues with model export, pre-processing input data, or configuring the inference code.

Let's diagnose the issue step by step:
1. **Verify ONNX model**: Ensure the ONNX model was exported correctly and is not corrupted. Try to run inference with this ONNX model in a Python environment to confirm that it yields non-zero bounding boxes.

2. **Check input preprocessing**: Review the image preprocessing steps in your C++ inference code. The input size and normalization should match the pre-processing used during model training and export.

3. **Inspect post-processing**: After getting the raw output from the model, the post-processing steps (e.g., non-maximum suppression) are crucial for obtaining the final bounding boxes. Ensure this part of your code is functioning as intended.

4. **Dependencies version**: Confirm that all dependencies, particularly the ONNX runtime and any other libraries used in the C++ inference code, are compatible with the exported model and perform as expected.

5. **Model input and output nodes**: Make sure you're accessing the correct input and output nodes for the ONNX model, as mismatches here could produce unexpected outcomes.

6. **Debug the C++ inference code**: Double-check your C++ inference code, especially parts you've modified. Potential bugs or misconfigurations could affect the resulting bounding boxes.

If after performing these checks the issue remains unresolved, I suggest you isolate a minimal code snippet that reproduces the issue and provide it within the GitHub Issues section for further assistance. When submitting additional details, please ensure your explanation remains concise and focuses on the problem at hand. Moreover, submitting a pull request with a proposed fix, if you've identified one, is always appreciated by the community.

Keep us updated on your progress, and we'll be glad to assist you further. 👍

**Comment by zp870420** [comment]
@glenn-jocher @all-for-code I have the same question.you are right "Check if the ONNX model has any specific layers or operations that might not be fully compatible with the OpenCV CUDA backend." 。I debug ，I find two operations dont run on cuda
![屏幕截图 2023-12-03 145442](https://github.com/ultralytics/ultralytics/assets/6006799/a56a4e55-5375-49f7-aeaa-97a49603206f)
![屏幕截图 2023-12-03 145508](https://github.com/ultralytics/ultralytics/assets/6006799/104ad0bc-cf07-4a51-af92-0839c9e9309f)
![屏幕截图 2023-12-03 145529](https://github.com/ultralytics/ultralytics/assets/6006799/932fb5f3-10bf-43eb-82d4-5011c5614848)

**Comment by all-for-code** [comment]
> @glenn-jocher @all-for-code I have the same question.you are right "Check if the ONNX model has any specific layers or operations that might not be fully compatible with the OpenCV CUDA backend." 。I debug ，I find two operations dont run on cuda ![屏幕截图 2023-12-03 145442](https://private-user-images.githubusercontent.com/6006799/287503361-a56a4e55-5375-49f7-aeaa-97a49603206f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE3NjMzMzAsIm5iZiI6MTcwMTc2MzAzMCwicGF0aCI6Ii82MDA2Nzk5LzI4NzUwMzM2MS1hNTZhNGU1NS01Mzc1LTQ5ZjctYWVhYS05N2E0OTYwMzIwNmYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMDVUMDc1NzEwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWI5ZTcwYTI3YzdmNjVhZjdjYjRjMzg4ZjFkNGQ5ODA3MGI1YTE0MmYyMDBiNjVjNGIzMTZiMjA3ZTRmZmU1ZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Ip70f30k7JFqpmxP8Ul1kCD79txJ_jBGq5pBDSpKD9s) ![屏幕截图 2023-12-03 145508](https://private-user-images.githubusercontent.com/6006799/287503368-104ad0bc-cf07-4a51-af92-0839c9e9309f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE3NjMzMzAsIm5iZiI6MTcwMTc2MzAzMCwicGF0aCI6Ii82MDA2Nzk5LzI4NzUwMzM2OC0xMDRhZDBiYy1jZjA3LTRhNTEtYWY5Mi0wODM5YzllOTMwOWYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMDVUMDc1NzEwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGZhNzk0ODRkNDg1ZThhNzJhOTc2YmQwYmE1OGEzMGQwYTE0ZDViYzNhODBmN2M1NTYwZjM2YjFlOGI4MjVlNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Ij0icJwwmNXJGsaehYT7kYcvkOQlqhP8HEtSQ4TMlPk) ![屏幕截图 2023-12-03 145529](https://private-user-images.githubusercontent.com/6006799/287503373-932fb5f3-10bf-43eb-82d4-5011c5614848.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE3NjMzMzAsIm5iZiI6MTcwMTc2MzAzMCwicGF0aCI6Ii82MDA2Nzk5LzI4NzUwMzM3My05MzJmYjVmMy0xMGJmLTQzZWItODJkNC01MDExYzU2MTQ4NDgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMDVUMDc1NzEwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NzI5ZTRhMmNhOWQxNDM2YjY0ZmZkZTIyNTdhY2YxMzlhNDc2OGRkYjNlN2IyODhhZjA5MzA1MTk5ZGM3ZTRiMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.tVwRsxBtOuUSdgijZa0K41O9kF2Rfo8xn3yGJbG8UWA)

@zp870420 @glenn-jocher  Hi, Thanks for your efforts in debugging the issue. I also found that running on the cpu was fine, but not with the gpu. 
You can set the runOnGPU variable in the main function to false and the result will be correct
![image](https://github.com/ultralytics/ultralytics/assets/110803332/f341c7ef-2961-4e9d-aed4-4ce220fab159)
Therefore, this problem may be caused by incomplete support for opencv4.8。
If you have further analysis results, please contact me in time, thank you very much！

**Comment by glenn-jocher** [comment]
@all-for-code Hi, it looks like you've correctly identified operations within the YOLOv8 ONNX model that aren’t fully supported by OpenCV's CUDA backend. Running these on the CPU gives you correct results, which indeed suggests a compatibility issue with the GPU.

Running the ONNX model on the CPU provides a temporary fix but isn't ideal for performance. You might want to either update OpenCV to the latest version with potentially improved CUDA support or switch to using the ONNX Runtime for GPU inference, which typically has broader compatibility with ONNX models.

Hope this helps to clarify your next steps. Keep up the good work, and please do reach out if you make any further progress or need additional assistance.

**Comment by glenn-jocher** [comment]
@zp870420 hi there, if the ONNX model isn't working with onnxruntime-gpu either, it's possible the issue may lie within the model export process or the ONNX model itself. Please make sure you:

1. Confirm that the onnxruntime-gpu is installed correctly and supports your GPU.
2. Double-check the export process for any potential issues.
3. Ensure your environment's dependencies are all compatible with the ONNX model you're using.

If bugs persist, consider opening an issue on the ONNX Runtime repository with details of the error for further assistance. 🛠️

**Comment by all-for-code** [comment]
> @all-for-code Hi, it looks like you've correctly identified operations within the YOLOv8 ONNX model that aren’t fully supported by OpenCV's CUDA backend. Running these on the CPU gives you correct results, which indeed suggests a compatibility issue with the GPU.
> 
> Running the ONNX model on the CPU provides a temporary fix but isn't ideal for performance. You might want to either update OpenCV to the latest version with potentially improved CUDA support or switch to using the ONNX Runtime for GPU inference, which typically has broader compatibility with ONNX models.
> 
> Hope this helps to clarify your next steps. Keep up the good work, and please do reach out if you make any further progress or need additional assistance.

Hi, @glenn-jocher
I think I should find the cause of the problem. When cuda inference is used, the picture in cv::Mat format obtained by cv.imread() should be converted to cv::cuda::GpuMat format. The code should look something like this:
![image](https://github.com/ultralytics/ultralytics/assets/110803332/ed62a777-dd74-48e2-ae79-f836fcaa2c40)
But I do not have the complete implementation code at present, if it is feasible, I will submit pr later

**Comment by glenn-jocher** [comment]
Hi @all-for-code,

Thank you for identifying the potential source of the problem! Indeed, ensuring that the image data is correctly transferred to the GPU as `cv::cuda::GpuMat` before running inference on the GPU is crucial. That step can often be overlooked and lead to such issues.

Please proceed with your proposed changes, and if they resolve the problem, a PR would be greatly appreciated by the community. Good luck, and we look forward to your contribution! 🌟
